File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res, curSet = [], []
        
        def dfs(i, nums, res, curSet):
            if i >= len(nums):
                res.append(curSet.copy())
                return
            
            curSet.append(nums[i])
            dfs(i + 1, nums, res, curSet)
            curSet.pop()
            dfs(i + 1, nums, res, curSet)

        dfs(0, nums, res, curSet)
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let res = []
        let curr = []

        const dfs = (i, total) => {
            if (total > target || i >= nums.length) {
                return
            }
            if (total === target) {
                res.push([...curr])
                return
            }

            curr.push(nums[i])
            dfs(i, total + nums[i])

            curr.pop()
            dfs(i + 1, total)
        }

        dfs(0, 0)

        return res
    }
}
